Add multi-registry documentation with schema v2 format (#1190)
open-telemetry/weaver: docs/multi-registry.md

The `dependencies` section lists other registries that your registry depends on. Each dependency has:

- **name**: A short name you'll use to reference this dependency
- **registry_path**: Path to the dependency's registry directory

The `registry_path` can be:

- A relative path: `data/multi-registry/otel_registry`
- An absolute path: `/path/to/otel/registry`
- A URL with optional archive path: `https://github.com/open-telemetry/semantic-conventions/archive/refs/tags/v1.37.0.zip[model]`

> **Note**: Currently, Weaver supports zero or one direct dependency per registry. However, transitive dependencies (dependencies of dependencies) are fully supported, allowing you to create multi-level registry hierarchies. See [issue #604](https://github.com/open-telemetry/weaver/issues/604) for tracking multiple direct dependencies support.

### Example: Three-Level Registry Hierarchy

Here's a practical example showing how registries can form a dependency chain:

**OTEL Registry** (`otel_registry/registry_manifest.yaml`):
```yaml
name: otel
description: OpenTelemetry base definitions
version: 1.0.0
repository_url: https://github.com/open-telemetry/semantic-conventions

> resolved_schema_url: https://opentelemetry.io/schemas/1.42.0

Reviewer: @copilot I was wrong, it should be

```suggestion
resolved_schema_url: https://opentelemetry.io/schemas/1.42.0/resolved_schema.yaml
```

**File**: `model/attributes.yaml`

```yaml
version: "2"
attributes:
  - key: example.message
    type: string
    brief: A simple message
    stability: development
    examples: ["Hello, World!"]
```

### 3. Define Custom Signals

**File**: `model/signals.yaml`

```yaml
version: "2"
spans:
  - name: example_message
    stability: development
    brief: This span represents a simple message.
    span_kind: client
    attributes:
      - ref: example.message
        requirement_level: required
      - ref: host.name         # Reference from OTel
        requirement_level: required
      - ref: host.arch         # Reference from OTel
        requirement_level: required

metrics:
  - name: example.counter
    stability: development
    brief: A counter of messages processed.
    instrument: counter
    unit: "1"
    attributes:
      - ref: host.name
        requirement_level: required
      - ref: host.arch
        requirement_level: required
```

### 4. Resolve the Schema

```bash
# Minimal schema (only referenced definitions)
weaver registry resolve model/

# Full schema (all definitions from OTel)
weaver registry resolve --include-unreferenced model/
```

## Common Use Cases

> ### Use Case 1: Vendor Extensions

Reviewer: @copilot these examples seem to repeat what'd already explained in the doc, maybe just list a few bullet points in the beginning of the doc, no need to repeat yaml again?

### Issue: Missing References

**Error**: Attribute `foo.bar` not found

**Causes**:
1. The attribute is not defined in any dependency
2. The signal containing the attribute is not imported
3. The attribute exists but the signal is garbage collected

**Solutions**:
- Add the attribute to an `imports` section
- Use `--include-unreferenced` to verify the attribute exists
- Check that the dependency is correctly declared in `registry_manifest.yaml`

### Issue: URL-based Dependencies Not Loading

**Error**: Cannot load registry from URL

**Solutions**:
- Verify the URL is accessible
- Check that archive paths (in `[brackets]`) point to the correct directory inside the archive
- Ensure the URL points to a registry with a valid `registry_manifest.yaml`

## Command Reference

All Weaver commands that work with registries support the `--include-unreferenced` flag:

```bash
# Resolve a registry
weaver registry resolve [--include-unreferenced] <registry-path>

# Generate code from a registry
weaver registry generate [--include-unreferenced] <registry-path> <template>

# Validate a registry

> weaver registry check [--include-unreferenced] <registry-path>

Reviewer: does it even make sense with --include-unreferenced?

# Update resolved schema
weaver registry update-markdown [--include-unreferenced] <registry-path>

# Calculate statistics on a registry
weaver registry stats [--include-unreferenced] <registry-path>

# Generate diff between registry versions

> weaver registry diff [--include-unreferenced] <baseline-path> <head-path>

Reviewer: does it make sense with --include-unreferenced?

- Build layered telemetry architectures with clear dependency chains

## Overview

A **multi-registry** setup consists of:

1. **Base registries** - Upstream registries that provide foundational definitions (attributes, metrics, events, entities)
2. **Dependent registries** - Your custom registries that build upon and reference definitions from base registries
3. **Registry manifest** - A `registry_manifest.yaml` file that declares dependencies between registries
4. **Imports** - Declarations in your schema files that specify which groups from dependencies you want to include

## Registry Manifest

Each registry must have a `registry_manifest.yaml` file in its root directory. This file defines the registry's identity and its dependencies on other registries.

### Basic Structure

```yaml
name: my-registry
description: My custom telemetry schema
version: 1.0.0
repository_url: https://example.com/schemas/
dependencies:
  - name: otel
    registry_path: path/to/otel/registry
```

### Required Fields

- **name**: Unique identifier for your registry
- **version**: Semantic version of your registry (also called `semconv_version`)

> - **repository_url**: Base URL where schema files are hosted (also called `schema_base_url`)

Reviewer: ```suggestion
- **repository_url**: Base URL where schema files are hosted
```

not sure if `schema_base_url` is useful for external readers

### Declaring Dependencies

The `dependencies` section lists other registries that your registry depends on. Each dependency has:

- **name**: A short name you'll use to reference this dependency
- **registry_path**: Path to the dependency's registry directory

The `registry_path` can be:

- A relative path: `data/multi-registry/otel_registry`
- An absolute path: `/path/to/otel/registry`
- A URL with optional archive path: `https://github.com/open-telemetry/semantic-conventions/archive/refs/tags/v1.37.0.zip[model]`

> **Note**: Currently, Weaver supports zero or one direct dependency per registry. However, transitive dependencies (dependencies of dependencies) are fully supported, allowing you to create multi-level registry hierarchies. See [issue #604](https://github.com/open-telemetry/weaver/issues/604) for tracking multiple direct dependencies support.

### Example: Three-Level Registry Hierarchy

Here's a practical example showing how registries can form a dependency chain:

**OTEL Registry** (`otel_registry/registry_manifest.yaml`):
```yaml
name: otel
description: OpenTelemetry base definitions
version: 1.0.0

> repository_url: https://opentelemetry.io/schemas/

Reviewer: ```suggestion
repository_url: https://github.com/open-telemetry/semantic-conventions
resolved_schema_url: https://opentelemetry.io/schemas/1.42.0
```

(more realistic)

```

**Vendor Registry** (`acme_registry/registry_manifest.yaml`):
```yaml
name: acme
description: Acme vendor-specific definitions
version: 0.1.0
repository_url: https://acme.com/schemas/
dependencies:
  - name: otel

>     registry_path: ../otel_registry

Reviewer: would it be a realistic/recommended use case? 

```suggestion
    registry_path: https://opentelemetry.io/schemas/1.42.0
```

In this setup:
- The `app` registry depends on `acme`
- The `acme` registry depends on `otel`
- The `app` registry can use definitions from both `acme` and `otel` (transitive dependencies are supported)

## Using Imports

The `imports` section in your schema files specifies which groups from dependent registries you want to include. This allows you to selectively pull in only the definitions you need.

### Import Syntax

Add an `imports` section at the top level of your YAML schema file (alongside `groups`):

```yaml
groups:
  - id: my.custom.group
    type: attribute_group
    # ... group definition ...

imports:
  metrics:
    - example.*
  events:
    - session.start
  entities:
    - gcp.*
```

### Import Categories

Imports are organized by signal type:

- **metrics**: Import metric definitions
- **events**: Import event definitions
- **entities**: Import entity definitions

### Wildcard Patterns

You can use wildcards to import multiple groups at once:

- `example.*` - Imports all groups with names starting with `example.`
- `gcp.*` - Imports all groups with names starting with `gcp.`
- `session.start` - Imports only the specific `session.start` group

### How Imports Work

> When you import a group:

Reviewer: ```suggestion
When you import a signal definition:
```

> 1. The group definition from the dependency is included in your resolved schema

Reviewer: @copilot "group" is so v1, use attribute or signal definition depending on the context

      - ref: http.request.method  # From imported http metrics
      - ref: app.custom.field     # Custom attribute
```

### Use Case 3: Testing Registries

When testing, you can create isolated test registries:

```yaml
# test_data/test_registry/registry_manifest.yaml
name: test
version: 0.1.0
repository_url: https://test.local/
dependencies:
  - name: base
    registry_path: ../base_registry
```

## Best Practices

1. **Version Dependencies Explicitly**: Use specific version tags in URLs rather than `main` or `latest`
   ```yaml
   registry_path: https://github.com/org/repo/archive/refs/tags/v1.2.3.zip[model]
   ```

2. **Use Wildcard Imports Judiciously**: Prefer specific imports over wildcards when you know exactly what you need
   ```yaml
   # Specific - better for minimal schemas
   imports:
     events:
       - session.start
       - session.end

   # Wildcard - better for exploratory work
   imports:
     events:
       - session.*
   ```

3. **Document Your Dependencies**: Add clear descriptions in your registry manifest
   ```yaml
   description: |
     Custom telemetry extending OTel v1.37.0.
     Adds vendor-specific attributes for auction systems.
   ```

> 4. **Test Both Modes**: Test your schema both with and without `--include-unreferenced`

Reviewer 2: @copilot Remove this guidance.  This guidance doesn't achieve the goal (verifying all dependencies are accessible).

### Wildcard Patterns

You can use wildcards to import multiple signals at once:

> - `example.*` - Imports all signals with names starting with `example.`

Reviewer: @copilot 

consider this (correct my english and update next points accordingly)
```suggestion
- `example.*` - Imports all definitions of corresponding signal with names starting with `example.`
```

imports:
  metrics:
    - example.*  # Imports example.counter metric from dependency
```

**Application Registry** (`app_registry/app_registry.yaml`):
```yaml
version: "2"
attributes:
  - key: app.name
    type: string
    brief: Name of the application.

> spans:

Reviewer: @copilot, where did this span come from? it's not quite useful here
